File: solutions/R/stage_2/stage2_ghost_audit.R
#!/usr/bin/env Rscript

# Get command-line arguments
args <- commandArgs(trailingOnly = TRUE)

# Check if dataset path is provided
if (length(args) < 1) {
  stop("Usage: Rscript stage2_ghost_audit.R <dataset_path>")
}

dataset_path <- args[1]

# Read the CSV file
data <- read.csv(dataset_path, stringsAsFactors = FALSE)

# Check if dataset is empty
if (nrow(data) == 0) {
  stop("Dataset is empty")
}

# Get bias_key from first row
bias_key <- as.integer(data$bias_key[1])

# Filter and calculate total
filtered <- data[
  data$status == "ACTIVE" & 
  as.integer(data$authentic) == 1 & 
  as.integer(data$priority) >= 4,
]

total <- sum(as.integer(filtered$units) * as.integer(filtered$multiplier))

# Calculate code
code <- (total + bias_key) %% 1000000

# Print with leading zeros (6 digits)
cat(sprintf("%06d\n", code))

File: solutions/python/stage_2/stage2_ghost_audit.py
import csv
import sys


if len(sys.argv) < 2:
    raise SystemExit("Usage: python stage2_ghost_audit.py <dataset_path>")


dataset_path = sys.argv[1]

with open(dataset_path, newline="") as handle:
    rows = list(csv.DictReader(handle))

if not rows:
    raise SystemExit("Dataset is empty")

bias_key = int(rows[0]["bias_key"])

total = 0
for row in rows:
    if (
        row["status"] == "ACTIVE"
        and int(row["authentic"]) == 1
        and int(row["priority"]) >= 4
    ):
        units = int(row["units"])
        multiplier = int(row["multiplier"])
        total += units * multiplier

code = (total + bias_key) % 1_000_000
print(f"{code:06d}")
